File: HW/HW5/HW 5_2025 data files/tutDec10_load_files.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


from pandas import read_excel
from printarray import printarray


# In[2]:


KDF = read_excel('stiffness_matrix.xlsx')
MDF = read_excel('mass_matrix.xlsx')
CDF = read_excel('damping_matrix.xlsx')


# In[3]:


K = KDF.values
M = MDF.values
C = CDF.values


# In[4]:


print('Upper left sub-half-matrix of K')
for ii in range(4):
    for jj in range(4):print('{:.15f}  '.format(K[ii,jj]),end="")

    print('')

print('\nLower right sub-half-matrix of K')
for ii in range(4,8):
    for jj in range(4,8):print('{:.15f}  '.format(K[ii,jj]),end="")

    print('')

print('\nUpper right sub-half-matrix of K')
for ii in range(4):
    for jj in range(4,8):print('{:.15f}  '.format(K[ii,jj]),end="")

    print('')


# In[5]:


shouldbezero8times8 = abs(K - K.T)
print('shouldbezero8times8:')
for ii in range(8):
    printarray(shouldbezero8times8[ii,:],'.10f')


# In[6]:


print('\nUpper left sub-half-matrix of M')
for ii in range(4):
    for jj in range(4):print('{:.15f}  '.format(M[ii,jj]),end="")

    print('')

print('\nLower right sub-half-matrix of M')
for ii in range(4,8):
    for jj in range(4,8):print('{:.15f}  '.format(M[ii,jj]),end="")

    print('')


print('\nUpper right sub-half-matrix of M')
for ii in range(4):
    for jj in range(4,8):print('{:.15f}  '.format(M[ii,jj]),end="")

    print('')


# In[7]:


shouldbezero8times8 = abs(M - M.T)
print('shouldbezero8times8:')
for ii in range(8):
    printarray(shouldbezero8times8[ii,:],'.10f')


# In[8]:


print('\nUpper left sub-half-matrix of C')
for ii in range(4):
    for jj in range(4):print('{:.15f}  '.format(C[ii,jj]),end="")

    print('')

print('\nLower right sub-half-matrix of C')
for ii in range(4,8):
    for jj in range(4,8):print('{:.15f}  '.format(C[ii,jj]),end="")

    print('')


print('\nUpper right sub-half-matrix of C')
for ii in range(4):
    for jj in range(4,8):print('{:.15f}  '.format(C[ii,jj]),end="")

    print('')


# In[9]:


shouldbezero8times8 = abs(C - C.T)
print('shouldbezero8times8:')
for ii in range(8):
    printarray(shouldbezero8times8[ii,:],'.10f')

import numpy as np
from scipy.linalg import eigh

# --- 下面是添加的计算部分 ---

# 1. Solve the Generalized Eigenvalue Problem: (K - w^2*M)v = 0
# 使用 scipy.linalg.eigh，它专门用于求解对称矩阵 (Symmetric Matrices) 的广义特征值问题
# eigenvalues (w^2) 会自动按从小到大排序
evals, evecs = eigh(K, M)

# 2. Calculate Natural Frequencies (rad/s)
# evals 里面存的是 frequencies 的平方 (omega^2)，所以要开根号
natural_frequencies = np.sqrt(evals)

print("\n-------------------------------------------")
print("Results for Homework 5")
print("-------------------------------------------")

# 打印所有的 Natural Frequencies (作业第1部分)
print("Undamped Natural Frequencies (rad/s):")
for i, freq in enumerate(natural_frequencies):
    print(f"Mode {i+1}: {freq:.4f} rad/s")

# 3. Calculate the ratio for the lowest natural frequency (作业第2部分)
# eigh 返回的结果通常已经排序，第0个就是 lowest frequency
lowest_freq_idx = 0
lowest_mode_shape = evecs[:, lowest_freq_idx]

# 获取第一个位移 (index 0) 和最后一个位移 (index 7，因为是8自由度)
disp_first = lowest_mode_shape[0]
disp_last = lowest_mode_shape[7]

ratio = disp_first / disp_last

print(f"\nLowest Natural Frequency is: {natural_frequencies[lowest_freq_idx]:.4f} rad/s")
print(f"Displacement Ratio (1st / 8th) for this mode: {ratio:.6f}")

File: HW/HW5/HW 5_2025 data files/tutDec10_load_files.m
clearvars, clc

M = readmatrix("mass_matrixMATLAB.xlsx");
K = readmatrix("stiffness_matrixMATLAB.xlsx");
C = readmatrix("damping_matrixMATLAB.xlsx");

format long
disp('Upper left sub-half-matrix of K')
K(1:4,1:4)
disp('Lower right sub-half-matrix of K')
K(5:8,5:8)
disp('Upper right sub-half-matrix of K')
K(1:4,5:8)

shouldbezero8times8 = abs(K - transpose(K));
disp('shouldbezero8times8:')
for ii = 1:8
    fprintf(' %12.10f',shouldbezero8times8(ii,:))
    fprintf('\n')
end
fprintf('\n')

disp('Upper left sub-half-matrix of M')
M(1:4,1:4)
disp('Lower right sub-half-matrix of M')
M(5:8,5:8)
disp('Upper right sub-half-matrix of M')
M(1:4,5:8)

shouldbezero8times8 = abs(M - transpose(M));
disp('shouldbezero8times8:')
for ii = 1:8
    fprintf(' %12.10f',shouldbezero8times8(ii,:))
    fprintf('\n')
end
fprintf('\n')

disp('Upper left sub-half-matrix of C')
C(1:4,1:4)
disp('Lower right sub-half-matrix of C')
C(5:8,5:8)
disp('Upper right sub-half-matrix of C')
C(1:4,5:8)

shouldbezero8times8 = abs(C - transpose(C));
disp('shouldbezero8times8:')
for ii = 1:8
    fprintf(' %12.10f',shouldbezero8times8(ii,:))
    fprintf('\n')
end
fprintf('\n')

% ========================================================
%  以下是添加的计算部分 (Calculation Part)
% ========================================================

% 1. 求解广义特征值问题 (Generalized Eigenvalue Problem)
% 求解方程: (K - lambda * M) * v = 0
% [V, D] = eig(K, M)
% V 是特征向量矩阵 (每一列是一个 Mode Shape)
% D 是对角矩阵，对角线上的元素是特征值 (omega^2)
[V, D] = eig(K, M);

% 2. 提取自然频率 (Natural Frequencies)
% 从对角矩阵提取特征值
eigenvalues = diag(D); 

% 特征值是频率的平方，所以要开根号得到 omega (rad/s)
omega_n = sqrt(eigenvalues);

% 3. 排序 (Sorting)
% 虽然 eig 通常会排序，但为了保险起见，我们手动按从小到大排序
[omega_sorted, sort_idx] = sort(omega_n);
V_sorted = V(:, sort_idx); % 同时要把对应的特征向量(Mode Shapes)也排好序

% 打印结果：自然频率
disp('Undamped Natural Frequencies (rad/s):');
disp(omega_sorted);

% 4. 计算作业要求的比值 (Ratio Calculation)
% 找到最低频率对应的 Mode Shape (排序后的第1列)
lowest_mode_shape = V_sorted(:, 1);

% 获取第1个位移 (Index 1) 和最后一个位移 (Index 8 或 end)
disp_first = lowest_mode_shape(1);
disp_last = lowest_mode_shape(end);

% 计算比值
ratio = disp_first / disp_last;

fprintf('\n=== Results for Homework 5 ===\n');
fprintf('Lowest Natural Frequency: %.5f rad/s\n', omega_sorted(1));
fprintf('Highest Natural Frequency: %.5f rad/s\n', omega_sorted(8));

fprintf('Displacement Ratio (1st / 8th): %.6f\n', ratio);